osu.Game/Graphics/UserInterface/BarGraph.cs

```diff
@@ -5,77 +5,233 @@
+
+            public override void ApplyState()
+            {
+                base.ApplyState();
+
+                shader = Source.shader;
+                texture = Source.texture;
+                drawSize = Source.DrawSize;
+                direction = Source.direction;
+
+                bars.Clear();
+                bars.AddRange(Source.bars);
+            }
+
+            public override void Draw(IRenderer renderer)
+            {
+                base.Draw(renderer);
+
+                if (!bars.Any())
+                    return;
+
+                shader.Bind();
+
+                for (int i = 0; i < bars.Count; i++)
+                {
+                    var bar = bars[i];
+
+                    float barHeight = drawSize.Y * ((direction == BarDirection.TopToBottom || direction == BarDirection.BottomToTop) ? bar.IntermediateValue : bar.ShortSide);
+                    float barWidth = drawSize.X * ((direction == BarDirection.LeftToRight || direction == BarDirection.RightToLeft) ? bar.IntermediateValue : bar.ShortSide);
+
+                    Vector2 topLeft;
+
+                    switch (direction)
+                    {
+                        default:
+                        case BarDirection.LeftToRight:
+                            topLeft = new Vector2(0, i * barHeight);
+                            break;
+
+                        case BarDirection.RightToLeft:
+                            topLeft = new Vector2(drawSize.X - barWidth, i * barHeight);
+                            break;
+
+                        case BarDirection.TopToBottom:
+                            topLeft = new Vector2(i * barWidth, 0);
+                            break;
+
+                        case BarDirection.BottomToTop:
+                            topLeft = new Vector2(i * barWidth, drawSize.Y - barHeight);
+                            break;
+                    }
+
+                    Vector2 topRight = topLeft + new Vector2(barWidth, 0);
+                    Vector2 bottomLeft = topLeft + new Vector2(0, barHeight);
+                    Vector2 bottomRight = bottomLeft + new Vector2(barWidth, 0);
+
+                    var drawQuad = new Quad(
+                        Vector2Extensions.Transform(topLeft, DrawInfo.Matrix),
+                        Vector2Extensions.Transform(topRight, DrawInfo.Matrix),
+                        Vector2Extensions.Transform(bottomLeft, DrawInfo.Matrix),
+                        Vector2Extensions.Transform(bottomRight, DrawInfo.Matrix)
+                    );
+
+                    renderer.DrawQuad(texture, drawQuad, DrawColourInfo.Colour);
+                }
+
+                shader.Unbind();
+            }
+        }
+
+        private struct BarStruct
+        {
+            public float OldValue { get; set; }
+            public float Value { get; set; }
+            public float IntermediateValue { get; set; }
+            public float ShortSide { get; set; }
```

Comment:
Not a fan of this naming. If it were up to me:

* `BarStruct` -> `BarInfo`
* `OldValue` -> `InitialLength`
* `Value` -> `FinalLength`
* `IntermediateValue` -> `InstantaneousLength`
* `ShortSide` -> `Breadth`

That said this structure may be changing quite a bit (see comments above re: array of structs vs struct of arrays, and the `ShortSide` being repeated for every bar for not much reason).